staking/staking-specification.md

```diff
@@ -0,0 +1,815 @@
+
+### 4.1 Ending One Epoch, and Starting a New One
+
+A new epoch automatically begins when the current epoch ends. Anyone can "end" an epoch by calling the staking contract after the minimum epoch period has elapsed.
+
+```
+function endEpoch()
+    external
+    returns (uint256 poolsRemaining);
+```
+
+Note: The return value (`poolsRemaining`) is described in detail in Section 7 Liquidity Rewards.
+
+
+## 5 Staking
+
+ZRX is staked by depositing tokens into the Staking Contract. The diagram below illustrates this workflow.
+
+<p align="center"><img src="./assets/Staking.png" width="700" /></p>
+
+ZRX can simlarly be unstaked by withdrawing tokens from the Staking contract. However, there are time-restrictions on unstaking ZRX, which are discussed more later in this section.
+
+Below is the interface for staking and unstaking.
+
+```
+/// @dev Stake ZRX tokens. Tokens are deposited into the ZRX Vault.
+/// @param amount of ZRX to stake.
+function stake(uint256 amount) external;
+
+/// @dev Unstake. Tokens are withdrawn from the ZRX Vault and returned to the owner.
+/// @param amount of ZRX to unstake.
+function unstake(uint256 amount) external;
+```
+
+### 5.1 Staking Pools
+
+Staking pools can be created to leverage the weight of other stakers. A pool has a single operator and any number of members, who delegate their ZRX to the pool. Any staker can create a pool, although at present it is only beneficial for market makers to create staking pools. This is discussed more in Section 6, along with details on creating a staking pool.
+
+|Term|Definition  |
+|--|--|
+| Pool Id | A unique id generated by this contract and assigned to each pool when it is created. |
+| Pool Operator | The creator and operator of the pool. |
+| Pool Members | Members of the pool who opted-in by delegating to the pool. |
+
+
+### 5.2 Stake Status
+Each staked ZRX has an associated status that reflects its utility within the 0x Ecosystem.
+
+|Status|Definition  |
+|--|--|
+| Active | Can be used to participate in governance. This is the default status. |
+| Inactive | Carries no utility within the 0x ecosystem. |
+| Inactive & Withdrawable | Once ZRX has been inactive for one full epoch it can be withdrawn. |
+| Delegated | ZRX is delegated to a pool; can be used in governance + contribute to liquidity rewards. |
+
+There is a single function for moving stake between statuses:
+```
+/// @dev Moves stake between statuses: 'active', 'inactive' or 'delegated'.
+///      This change comes into effect next epoch.
+/// @param from status to move stake out of.
+/// @param to status to move stake into.
+/// @param amount of stake to move.
+function moveStake(
+    IStructs.StakeInfo calldata from,
+    IStructs.StakeInfo calldata to,
+    uint256 amount
+) external;
+
+/// @dev Info used to describe a status.
+/// @param status of the stake.
+/// @param poolId Unique Id of pool. This is set when status=DELEGATED.
+struct StakeInfo {
```

Comment:
Since this struct is included here, I think that it would be a bit more readable to change `IStructs.StakeInfo` to just `StakeInfo` in the parameter.

Reply:
Also, I liked how you prefaced the getters in the section below by the struct definition. What do you think about doing that here with the struct and enum?